### PR TITLE
Problem: go1.20 is not used

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #844
Solution:
- use go 1.20
- update nixpkgs to master, and remove custom rocksdb expression in favor of builtin one.
- setup go coverage support for integration test
- don't upload the coverage reports yet due to a bug

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -52,6 +52,17 @@ jobs:
...
           if-no-files-found: ignore
+      - name: Convert coverage data
+        run: |
+          nix profile install nixpkgs#go_1_20
+          go tool covdata textfmt -i=coverage -o profile.txt
+      - name: Upload coverage report
+        uses: codecov/codecov-action@v3
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          file: ./coverage.*.txt
```
Comment (yihuang) [inline .github/workflows/test.yml:63]:
```suggestion
```

File: default.nix [hunk 2]
```diff
@@ -34,6 +35,7 @@ buildGoApplication rec {
   modules = ./gomod2nix.toml;
   pwd = src; # needed to support replace
   subPackages = [ "cmd/cronosd" ];
+  buildFlags = lib.optionalString coverage "-cover";
```
Comment (tomtau) [inline default.nix:38]: this shows warnings in gomod2nix -- probably a todo to fix when it's possible to directly configure it there?

Comment (yihuang) [reply default.nix:38]: yeah, I just opened an issue there: https://github.com/nix-community/gomod2nix/issues/104

**Review by mmsqe — APPROVED** [review]

**Comment by yihuang** [comment]
```
$ go tool covdata textfmt -i=integration_tests/coverage -o profile.txt
error: error: reading counter data file integration_tests/coverage/covcounters.627546bfc793cab3f06764545dba1bd1.91113.1675846217322297000: error: short read on string table
```

I can reproduce this issue, maybe something wrong with this new golang feature.

**Comment by mmsqe** [comment]
> go tool covdata textfmt -i=integration_tests/coverage -o profile.txt

I thought it error on report on 1st time, since profile.txt is normal when I tried
```
integration_tests/coverage/
profile.txt
```

**Comment by yihuang** [comment]
I guess we should merge the nixpkgs and golang version upgrade first, then debugging the coverage thing, otherwise, the integration test workflow takes 2hours each time.

**Review by JayT106 — APPROVED** [review]
LGTM, could you describe the purpose we change the nixpkgs from release version to the master? Thanks.

**Comment by yihuang** [comment]
> LGTM, could you describe the purpose we change the nixpkgs from release version to the master? Thanks.

just for the go 1.20 which is only in the master branch.

**Comment by yihuang** [comment]
https://github.com/golang/go/issues/58411
it's indeed a bug in golang runtime, so let's update golang only, without using the `-cover` for now.

**Pull request merged**